jmun/embeddings-modal: initialized embeddings and modal

Jumshim opened the pull request:
Familiarized myself with embedding models -- stepping stone before hosting them on Modal

Jumshim committed:
jmun/embeddings-modal: initialized embeddings

Declared pydantic types for EmbeddingInput,
Embedding, Usage, and EmbeddingOutput. Also
initialized a template for the FastAPI
endpoints to retrieve embeddings from
Modal and BaseTen

jxnl commented on applications/embeddings-modal/main.py:
```diff
@@ -0,0 +1,50 @@
+import torch
+from transformers import AutoModel, AutoTokenizer
+
+base_name = "BAAI/bge-base-en-v1.5"
```
Use a dict

jxnl commented on applications/embeddings-modal/main.py:
```diff
@@ -0,0 +1,50 @@
+    self.model = AutoModel.from_pretrained(model_name)
+    self.model.eval()
+
+  def get_embedding(self, text):
```
Also check out Justin's pr

Jumshim replied:
ooh gotcha. I'll use it as a reference in setting up modal and FastAPI, model the code to reflect OpenAI, and update the readme and stuff. I'll add the code to run benchmarks in a separate PR too.

Anything else I should take into account?

jxnl replied:
nope, hows the progress with baseten?

Jumshim committed:
jmun/embeddings-modal: initialized modal app

Established a modal app called 'embedder'
that's connected with the FastAPI app in
`main.py`.

Jumshim committed:
jmun/embeddings-modal: README and Dockerfile

Updated the README and Dockerfile templates
to be compatibile with this specific project.

'embeddings-modal'

Jumshim commented on applications/embeddings-modal/src/embeddings.py:
```diff
@@ -0,0 +1,55 @@
+  embeddings = model.encode(sentences, convert_to_tensor=True)
+  return EmbeddingOutput(
+    object= "list",
+    data=[Embedding(embedding=e, index=i) 
```
🐛 This line seems to bug out on Modal -- a type validation error occurs where `data=[Embedding(embedding=e, index=i` is not recognized as a list. 

I've even tried putting it all in the same line like:
`data=[Embedding(embedding=e, index=i) for i, e in enumerate(embeddings)],`, but the type error still arises.

Jumshim commented:
**11/27/2023 Update**

Back from Thanksgiving! Here are my updates since my last push:

1) Revised my implementation of Embeddings to match the other examples in `fastllm`.

2) Created the FastAPI Endpoint and tested it locally to ensure that it works

3) Connected the FastAPI endpoint to Modal and tested it out (note, this PR is not ready to be pushed because there are some issues with Modal. I noted the problems in a comment within the code)

TODO: Finish modal connection then work on BaseTen (I got approved and will probably add BaseTen support in another PR). 

Will rename the directory from `embeddings-modal` to `embeddings-benchmarks` since I have now received access to BaseTen.

Jumshim commented on applications/embeddings-modal/src/embeddings.py:
```diff
@@ -0,0 +1,55 @@
+from pydantic import BaseModel, Field
+from typing import List, Union
+
+model_names = {
```
A bit unsure on how to integrate the various model versions when we're passing in the model name through the request body. I've currently got the model name set to what the user passes in, but I'm also considering an alternative where the request body will pass in like

`"type": "base"`, `"type": "small"`, or `"type": "large"`

rather than `"model": "BAAI/bge-base-en-v1.5"`...

jxnl replied:
usually you'll just do 

model: ...

and then have a method liek 

`/list_models` that returns the list in case anyoen wants to know whats available

Jumshim committed:
baseten and modal initial setup

Jumshim committed:
list_models get endpoint

Jumshim committed:
Linear benchmarking

jxnl commented on applications/embeddings-modal/baseten_embedder/model/model.py:
```diff
@@ -0,0 +1,80 @@
+def get_model(model_name=model_names["base"]) -> SentenceTransformer:
+  return SentenceTransformer(model_name)
+
+def embeddings(req):
```
both these functions could have been imported from the same `src/model.py` code eh?

Jumshim replied:
yeah i think so -- when I was first making these files, the other pr hadn't been merged yet I believe. so i recreated it, and it helped me understand some python along the way

Jumshim replied:
Oh wait, I remember why I put this in now. I couldn't get Baseten to work with local import files, and this was just a quick patch for the meantime. Baseten also doesn't have a community/discord/slack where I could ask the questions appropriately

jxnl commented:
1. lets add some benchmarking server side
2. deploy and test baseten too
3. and see how much faster we can get it in the next 2 days.

Jumshim committed:
added batch processing and benchmarking on modal'

Jumshim commented:
Batch Processing made Modal rub so much faster

Currently working on BaseTen

Jumshim committed:
finished baseten benchmarking and batch requests

jxnl commented:
can you add some some benchmarks here?

jxnl commented on applications/embeddings-modal/baseten_embedder/dataset.py:
```diff
@@ -0,0 +1,82 @@
+# Benchmarking for Batch Requests
+def get_embeddings(texts: List[str]) -> List[List[float]]:
+  input_data = json.dumps({"input": texts, "model": MODEL_NAME})
+  truss_cmd = ["truss", "predict", "-d", input_data]
```
why can't we do this via post request?

jxnl commented on applications/embeddings-modal/baseten_embedder/dataset.py:
```diff
@@ -0,0 +1,82 @@
+
+def save_to_excel(data):
+  df = pd.dataframe(data)
+  df.to_excel(EXCEL_FILE, index=False)
```
just do a csv instead

jxnl commented on applications/embeddings-modal/dataset.py:
```diff
@@ -0,0 +1,93 @@
+  end = time.time()
+  save_to_excel(embeddings)
+
+  print(f"--------TOTAL TIME TO PROCESS-------- \n {end-start}s")
```
can you leave all the printouts in the pr?
